### PR TITLE
[Universal Print] Adding support for task definitions and triggers

### PR DESCRIPTION
**Pull request by braedenp-msft**

### CODE REVIEW

File: api-reference/beta/resources/printjob.md [hunk 1]
```diff
@@ -23,6 +23,7 @@ Represents a print job that has been queued for a printer.
 | [Create](../api/printer-post-jobs.md) | [printJob](printjob.md) | Create a new print job object. |
 | [Start](../api/printjob-startprintjob.md)|None|Start the print job.|
 | [Cancel](../api/printjob-cancelprintjob.md)|None|Cancel the print job.|
+| [List tasks](../api/printjob-list-tasks.md)|[printTask](printtask.md) collection|Retrieve a list of [printTasks](printtask.md) triggered by this print job.|
```
Comment (nilakhan) [inline api-reference/beta/resources/printjob.md:26]: We dont support listing tasks for printjob today.
though $expand=tasks is supported for printJob.

File: api-reference/beta/resources/printtask.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Represents a task that is executing or has been executed as a result of an a Universal Print event.
+
+## Methods
+
+| Method       | Return Type | Description |
+|:-------------|:------------|:------------|
+| [List (from printTaskDefintion)](../api/printtaskdefinition-list-tasks.md) | [printTask](printtask.md) | Get a list of tasks that have been created based on a particular printTaskDefinition. The list includes currently running tasks and recently completed tasks. |
+| [List (from printJob)](../api/printjob-list-tasks.md) | [printTask](printtask.md) | Get a list of tasks that have been created based on event(s) related to a particular printJob. The list includes currently running tasks and recently completed tasks. |
```
Comment (nilakhan) [inline api-reference/beta/resources/printtask.md:23]: redundant comment, we dont support list tasks from PrintJob today. we support expanding tasks from printJob.

File: api-reference/beta/resources/printtask.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+## Properties
+| Property     | Type        | Description |
+|:-------------|:------------|:------------|
+|id|String|The printTask's identifier. Read-only.|
+|status|[printTaskStatus](printtaskstatus.md)|The current execution status of this printTask. **The calling application is responsible for updating this status when processing is finished.** Failure to report completion will result in the related print job being blocked from printing and eventually deleted. |
```
Comment (nilakhan) [inline api-reference/beta/resources/printtask.md:29]: Exception to this is "redirecting print job" where the active task on original print job is automatically completed when the job is redirected.

File: api-reference/beta/resources/printtask.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+|parentUrl|String|The URL for the print entity that triggered this task. e.g., `https://graph.microsoft.com/beta/print/printers/{printerId}/jobs/{jobId}`. Read-only.|
+
+## Relationships
+| Relationship | Type        | Description |
+|:-------------|:------------|:------------|
+|trigger|[printTaskTrigger](printtasktrigger.md)|The printTaskTrigger triggered this task's execution. Read-only.|
```
Comment (nilakhan) [inline api-reference/beta/resources/printtask.md:35]: missing "that":

The printTaskTrigger that triggered this task's execution

File: api-reference/beta/resources/printtaskdefinition.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+doc_type: resourcePageType
+---
+
+# printTaskDefinition resource type
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Represents a task that can be triggered when various events occur within Universal Print. See the [Pull Print Overview](/graph/universal-print-concept-overview.md#extending-universal-print-to-support-pull-printing) to understand how you can use this API to add Pull Print support to Universal Print.
```
Comment (nilakhan) [inline api-reference/beta/resources/printtaskdefinition.md:16]: Suggestion: 
Represents an abstract definition for a task that can be triggered when various events occur within Universal Print.

File: api-reference/beta/resources/printtasktrigger.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+doc_type: resourcePageType
+---
+
+# printTaskTrigger resource type
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Determines the conditions under which a new [printTask](printtask.md) will be executed based on the associated [printTaskDefinition](printtaskdefinition.md).
```
Comment (nilakhan) [inline api-reference/beta/resources/printtasktrigger.md:16]: suggestion: 
conditions => condition, 
since taskTrigger contains single event.

File: api-reference/beta/resources/printtasktrigger.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+## Properties
+| Property     | Type        | Description |
+|:-------------|:------------|:------------|
+|id|String|The printTaskTrigger's identifier. Read-only.|
+|event|[printEvent](printevent.md)|The Universal Print event that will cause a new [printTask](printtask.md) to be executed.|
+
+## Relationships
+| Relationship | Type        | Description |
+|:-------------|:------------|:------------|
+|definition|[printTaskDefinition](printtaskdefinition.md)|The [printTaskDefinition](printtaskdefinition.md) that defines how a new [printTask](printTask.md) will be executed. Read-only.|
```
Comment (nilakhan) [inline api-reference/beta/resources/printtasktrigger.md:34]: suggestion: 
Abstract definition based on which the task will triggered.

I think task definition does not define how a task is executed.

File: api-reference/beta/api/print-list-taskdefinitions.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+doc_type: apiPageType
+---
+
+# List taskDefinitions
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Retrieve a list of **taskDefinitions** defined for the user or app's tenant.
```
Comment (nilakhan) [inline api-reference/beta/api/print-list-taskdefinitions.md:16]: Retrieve a list of **taskDefinitions** defined for the tenant by the requesting app.

File: api-reference/beta/api/print-update-taskdefinition.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+##### Response
+The following is an example of the response.
+>**Note:** The response object shown here might be shortened for readability. All the properties will be returned from an actual call.
+<!-- {
+  "blockType": "response",
+  "truncated": true,
+  "@odata.type": "microsoft.graph.printTaskDefinition"
+} -->
+```http
+HTTP/1.1 200 OK
```
Comment (nilakhan) [inline api-reference/beta/api/print-update-taskdefinition.md:83]: Response status will be 204 and response body will be empty

File: api-reference/beta/api/printer-list-tasktriggers.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+## Permissions
+One of the following permissions is required to call this API. To learn more, including how to choose permissions, see [Permissions](/graph/permissions-reference).
+
+In addition to the following permissions, the user's tenant must have an active Universal Print subscription.
+
+|Permission type | Permissions (from least to most privileged) |
+|:---------------|:--------------------------------------------|
+|Delegated (work or school account)| Not supported. |
+|Delegated (personal Microsoft account)|Not Supported.|
+|Application| Printer.Read.All, Printer.ReadWrite.All, Printer.FullControl.All |
```
Comment (nilakhan) [inline api-reference/beta/api/printer-list-tasktriggers.md:27]: This should be supported for delegated permissions: Printer.Read.All, Printer.ReadWrite.All, Printer.FullControl.All
Its not supported for application permissions

File: api-reference/beta/api/printer-list-tasktriggers.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+## HTTP request
+<!-- { "blockType": "ignored" } -->
+```http
+GET /print/printers/{id}/jobs
```
Comment (nilakhan) [inline api-reference/beta/api/printer-list-tasktriggers.md:32]: GET /print/printers/{id}/taskTriggers

File: api-reference/beta/api/printjob-redirect.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+```http
+POST /print/printers/{id}/jobs/{id}/redirect
+```
+## Request headers
+| Name          | Description   |
+|:--------------|:--------------|
+| Authorization | Bearer {token}. Required. |
+
+## Request body
+In the request body, supply the ID of the printer that the print job should be moved to.
```
Comment (nilakhan) [inline api-reference/beta/api/printjob-redirect.md:40]: moved => redirected

File: api-reference/beta/api/printjob-redirect.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+| Property     | Type        | Description |
+|:-------------|:------------|:------------|
+|destinationPrinterId|String|The ID of the printer the print job should be moved to.|
```
Comment (nilakhan) [inline api-reference/beta/api/printjob-redirect.md:44]: moved => redirected

File: api-reference/beta/api/printjob-redirect.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+## Response
+If successful, this method returns a `200 OK` response code and a [printJob](../resources/printjob.md) object in the response body.
```
Comment (nilakhan) [inline api-reference/beta/api/printjob-redirect.md:47]: If successful, this method returns a `200 OK` response code and a [printJob](../resources/printjob.md) object queued in the destination printer.

File: api-reference/beta/api/printer-post-tasktriggers.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+---
+
+# Create taskTrigger
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Create a new [task trigger](../resources/printtasktrigger.md) on the specified [printer](../resources/printer.md).
+
```
Comment (nilakhan) [inline api-reference/beta/api/printer-post-tasktriggers.md:17]: please also add that **currently** only one task trigger is supported per printer and **currently**, only the app that created(registered) the printer can add task trigger to it.

File: api-reference/beta/api/printtask-get.md [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+## Permissions
+One of the following permissions is required to call this API. To learn more, including how to choose permissions, see [Permissions](/graph/permissions-reference).
+
+In addition to the following permissions, the user's tenant must have an active Universal Print subscription.
+
+|Permission type | Permissions (from least to most privileged) |
+|:---------------|:--------------------------------------------|
+|Delegated (work or school account)| Users.Read.All |
+|Delegated (personal Microsoft account)|Not Supported.|
+|Application| Not supported. |
```
Comment (nilakhan) [inline api-reference/beta/api/printtask-get.md:29]: a) this API is supported only application permissions right now. 
b) this is a miss in service that we have not added permission checks for this, Can we still add in documentation that this requires Application permission scope, TaskDefinition.ReadWrite.All? We will fix it in service by next deployment

File: api-reference/beta/api/printtaskdefinition-update-task.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+##### Request
+The following is an example of the request.
+
+# [HTTP](#tab/http)
+<!-- {
+  "blockType": "request",
+  "name": "update_task"
+}-->
+```msgraph-interactive
+PATCH https://graph.microsoft.com/beta/print/taskDefinitions/3203656e-6069-4e10-8147-d25290b00a3c/tasks/d036638b-1272-4bba-9227-732463823ed3
```
Comment (nilakhan) [inline api-reference/beta/api/printtaskdefinition-update-task.md:56]: we need to add request body here.

File: api-reference/beta/resources/print.md [hunk 18]
```diff
@@ -25,6 +25,10 @@ When accompanied by a Universal Print subscription, the Print feature enables ma
 | [Create printerShare](../api/print-post-shares.md) | [printerShare](printershare.md) | Create a new printer share by posting to the **shares** collection. |
 | [Create printer](../api/printer-create.md) | [printerCreateOperation](printerCreateOperation.md) | Create (register) a new printer with Universal Print. |
 | [Update settings](../api/print-update-settings.md) |  [printSettings](printsettings.md) | Updates tenant-wide settings for the Universal Print service. |
+| [List taskDefinitions](../api/print-list-taskdefinitions.md) | [printTaskDefinition](printtaskdefinition.md) collection | Get a complete list of printTaskDefinitions created within Universal Print. |
```
Comment (nilakhan) [inline api-reference/beta/resources/print.md:28]: Get a tenant-wide list of printTaskDefinitions created within Universal Print

File: api-reference/beta/resources/printtaskdefinition.md [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+| [Create](../api/print-post-taskDefinitions.md) | [printTaskDefinition](printtaskdefinition.md) | Create a new printTaskDefinition. |
+| [Update](../api/print-update-taskDefinitions.md) | [printTaskDefinition](printtaskdefinition.md) | Update a printTaskDefinition. |
+| [Delete](../api/print-delete-taskDefinitions.md) | None | Delete a printTaskDefinition. |
+| [List tasks](../api/printtaskdefinition-list-tasks.md) | [printTask](printtask.md) | Get a list of tasks that have been created based on this definition. The list includes currently running tasks and recently completed tasks. |
+| [Get task](../api/printtask-get.md) | [printTask](printtask.md) | Gets a task that has been created based on this definition. |
+| [Update task](../api/printtaskdefinition-update-task.md) | None | Update a task that has been created based on this definition. **Applications that register task triggers are responsible for updating task status when processing is finished.** Failure to report completion will result in the related print job being blocked from printing and eventually deleted. |
```
Comment (nilakhan) [inline api-reference/beta/resources/printtaskdefinition.md:30]: Applications that register task triggers are responsible for updating task status when processing is finished, unless the related printJob has been redirected to another printer.

File: api-reference/beta/resources/printtaskstatus.md [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+doc_type: resourcePageType
+---
+
+# printTaskStatus complex type
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Represents the current execution status of a [printTask](printtask.md). **The calling application is responsible for updating this status when processing is finished.** Failure to report completion will result in the related print job being blocked from printing and eventually deleted.
```
Comment (nilakhan) [inline api-reference/beta/resources/printtaskstatus.md:16]: Applications that register task triggers are responsible for updating task status when processing is finished, unless the related printJob has been redirected to another printer.

File: api-reference/beta/resources/printtasktrigger.md [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+doc_type: resourcePageType
+---
+
+# printTaskTrigger resource type
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Determines the condition under which a new [printTask](printtask.md) will be executed based on the associated [printTaskDefinition](printtaskdefinition.md).
```
Comment (nilakhan) [inline api-reference/beta/resources/printtasktrigger.md:16]: executed => triggered

File: api-reference/beta/resources/printtasktrigger.md [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+| Method       | Return Type | Description |
+|:-------------|:------------|:------------|
+| [Get](../api/printtasktrigger-get.md) | [printTaskTrigger](printtasktrigger.md) | Get the printTaskTrigger associated with a particular [printTask](printtask.md). |
+| [List](../api/printer-list-tasktriggers.md) | [printTaskTrigger](printtasktrigger.md) collection | Get a list of printTaskTriggers associated with a particular [printer](printer.md). |
+
+## Properties
+| Property     | Type        | Description |
+|:-------------|:------------|:------------|
+|id|String|The printTaskTrigger's identifier. Read-only.|
+|event|[printEvent](printevent.md)|The Universal Print event that will cause a new [printTask](printtask.md) to be executed. Valid values are described in the following table.|
```
Comment (nilakhan) [inline api-reference/beta/resources/printtasktrigger.md:31]: executed => triggered

**Comment by braedenp-msft** [comment]
@Lauragra this one's ready to go 😊

File: concepts/universal-print-concept-overview.md [hunk 23]
```diff
@@ -57,6 +57,28 @@ Sharing a printer creates a [printerShare](/graph/api/resources/printershare?vie
...
+
+2. [Register one or more virtual printers](/graph/api/printer-create?view=graph-rest-beta) using an administrator authentication token and a `null` deviceId. A "virtual printer" is just a printer object in Universal Print without a physical device attached. Usually, users will print to virtual printers and later pick up their print jobs at a physical print device. See step 6.
+
+3. [Update the attributes of your printer](/graph/api/printer-update?view=graph-rest-beta) by using application permissions and an `application/ipp` media type (see examples).
+
+4. [Create a task trigger for your virtual printer](/graph/api/printer-post-tasktriggers?view=graph-rest-beta) that will generate a notification for your application when events occur.
+
+5. When a print job is submitted to the virtual printer, it will be paused due to the [printTaskTrigger](/graph/api/resources/printtasktrigger?view=graph-rest-beta). A [printTask](/graph/api/resources/printtask?view=graph-rest-beta) will be created based on the associated [printTaskDefinition](/graph/api/resources/printtaskdefinition?view=graph-rest-beta).
```
Comment (nilakhan) [inline concepts/universal-print-concept-overview.md:76]: A [printTask](/graph/api/resources/printtask?view=graph-rest-beta) with state "processing" will be created based on the associated [printTaskDefinition]

File: concepts/universal-print-concept-overview.md [hunk 24]
```diff
@@ -57,6 +57,28 @@ Sharing a printer creates a [printerShare](/graph/api/resources/printershare?vie
...
 
+### Extending Universal Print to support Pull Printing
+
+The Microsoft Graph Universal Print API enables your application to support Pull Printing (also known as Follow-Me or Find-Me print)! To set up Pull Printing, you will register triggers that notify your application (via service-to-service communication) when certain print events happen, such as a print job being started.
+
+These triggers enable your application to interrupt the print workflow to do things such as redirecting jobs to different printers and modifying the document payload.
+
+Follow these steps to enable Pull Printing:
+
+1. [Create a printTaskDefinition](/graph/api/print-post-taskdefinitions?view=graph-rest-beta) using application permissions. This abstract task definition will be used to create task instances that your service can use to tell Universal Print when you're done processing print jobs.
```
[the remark this answers is not in the document]
Comment (braedenp-msft) [reply concepts/universal-print-concept-overview.md:68]: Nice wording! I actually removed the last sentence altogether because tasks are completed automatically in the Pull Print scenario.

File: concepts/universal-print-concept-overview.md [hunk 25]
```diff
@@ -57,6 +57,28 @@ Sharing a printer creates a [printerShare](/graph/api/resources/printershare?vie
...
+
+2. [Register one or more virtual printers](/graph/api/printer-create?view=graph-rest-beta) using an administrator authentication token and a `null` deviceId. A "virtual printer" is just a printer object in Universal Print without a physical device attached. Usually, users will print to virtual printers and later pick up their print jobs at a physical print device. See step 6.
+
+3. [Update the attributes of your printer](/graph/api/printer-update?view=graph-rest-beta) by using application permissions and an `application/ipp` media type (see examples).
+
+4. [Create a task trigger for your virtual printer](/graph/api/printer-post-tasktriggers?view=graph-rest-beta) that will generate a notification for your application when events occur.
```
[the remark this answers is not in the document]
Comment (braedenp-msft) [reply concepts/universal-print-concept-overview.md:74]: Removed the bit about notifications since we don't support it yet. Can put that in overall workflow docs when that's ready 😊

File: api-reference/beta/resources/printtaskdefinition.md (hunk 19 above)
Comment (nilakhan) [reply api-reference/beta/resources/printtaskdefinition.md:30]: this one is pending

Comment (braedenp-msft) [reply api-reference/beta/resources/printtaskdefinition.md]: Fixed!

**Comment by braedenp-msft** [comment]
@Lauragra As per your request in email, this is a time-sensitive PR that a partner is waiting for to get them unblocked.

Can we please prioritize? 😊

File: api-reference/beta/api/print-delete-taskdefinition.md [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+
+# Delete taskDefinition
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Delete a **taskDefinition**.
+
+See the [Pull Print Overview](/graph/universal-print-concept-overview.md#extending-universal-print-to-support-pull-printing) to understand how you can use this API to add Pull Print support to Universal Print.
```
Comment (Lauragra) [inline api-reference/beta/api/print-delete-taskdefinition.md:18]: Global: Pull Printing is not a proper noun; please use lowercase for this term. 

For consistency with Microsoft style, please revise this sentence as follows: "For details about how to use this API to add pull printing support to Universal Print, see [Extending Universal Print to support pull printing](/graph/universal-print-concept-overview.md#extending-universal-print-to-support-pull-printing)."

File: api-reference/beta/api/print-delete-taskdefinition.md [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+| Authorization | Bearer {token}. Required. |
+
+## Request body
+Do not supply a request body for this method.
+
+## Response
+If successful, this method returns a `204 No Content` response code. It does not return anything in the response body.
+
+## Example
+##### Request
```
Comment (Lauragra) [inline api-reference/beta/api/print-delete-taskdefinition.md:48]: Global: These should be H3s (###). The new stub generator tool should output the correct heading levels.

File: api-reference/beta/api/print-post-taskdefinitions.md [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+## HTTP request
+<!-- { "blockType": "ignored" } -->
+```http
+POST /print/taskDefinitions
+```
+
+## Request headers
+| Name      |Description|
+|:----------|:----------|
+| Authorization | Bearer {token}. Required. |
```
Comment (Lauragra) [inline api-reference/beta/api/print-post-taskdefinitions.md:40]: Add the Content-type header to the Request headers table.

File: api-reference/beta/api/print-delete-taskdefinition.md [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+The following is an example of the request.
+
+# [HTTP](#tab/http)
```
Comment (Lauragra) [inline api-reference/beta/api/print-delete-taskdefinition.md:51]: Global: Please remove these H1s. They are only relevant when we add the snippets for the different languages, which happens via automation each Tuesday.

File: api-reference/beta/api/print-delete-taskdefinition.md [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+### Request
+The following is an example of the request.
+
+# [HTTP](#tab/http)
+<!-- {
+  "blockType": "request",
+  "name": "delete_taskdefinition"
+}-->
+```http
+DELETE /print/taskDefinitions/4c6a0f26-8e5d-4bf6-91e6-4a5731adec19
```
Comment (Lauragra) [inline api-reference/beta/api/print-delete-taskdefinition.md:57]: Global: Please make sure that your examples show the full URL.

File: api-reference/beta/api/print-delete-taskdefinition.md [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+
+# Delete taskDefinition
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Delete a **taskDefinition**.
+
+See the [pull print overview](/graph/universal-print-concept-overview.md#extending-universal-print-to-support-pull-printing) to understand how you can use this API to add pull print support to Universal Print.
```
Comment (Lauragra) [inline api-reference/beta/api/print-delete-taskdefinition.md:18]: For better readability, it's helpful to tell the reader why they should go to a new URL before you provide the URL. So this sentence should be restructured, per my earlier comment. "To learn how to use this API to add pull print support to Universal Print, see...". This is consistent with how we structure cross-references. Per the Writing Style Guide, we should be consistent in presentation. https://docs.microsoft.com/en-us/style-guide/scannable-content/

Comment (braedenp-msft) [reply api-reference/beta/api/print-delete-taskdefinition.md]: Apologies, forgot to change the phrasing per your earlier comment! Fixed now 😊

File: api-reference/beta/api/print-delete-taskdefinition.md [hunk 32]
```diff
@@ -0,0 +1,80 @@
...
+
+# Delete taskDefinition
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Delete a **taskDefinition**.
+
+For details about how to use this API to add pull printing support to Universal Print, see [Extending Universal Print to support pull printing](/graph/universal-print-concept-overview#extending-universal-print-to-support-pull-printing).
```
Comment (Lauragra) [inline api-reference/beta/api/print-delete-taskdefinition.md:18]: FYI, when I tested it on staging, this link didn't work. I think it will work when we remove the .md from the file name, but I need the build to finish in order to verify. If you get a chance, and you see that the build finishes, please verify the link and if it works, apply the link fix to all occurrences in the PR.

Comment (braedenp-msft) [reply api-reference/beta/api/print-delete-taskdefinition.md:18]: Will do, thanks for fixing!

Comment (braedenp-msft) [reply api-reference/beta/api/print-delete-taskdefinition.md:18]: Verified and updated the rest of the links!

File: concepts/universal-print-concept-overview.md [hunk 33]
```diff
@@ -57,6 +57,28 @@ Sharing a printer creates a [printerShare](/graph/api/resources/printershare?vie
...
+
+These triggers enable your application to interrupt the print workflow to do things such as redirecting jobs to different printers and modifying the document payload.
+
+Follow these steps to enable pull printing:
+
+1. [Create a printTaskDefinition](/graph/api/print-post-taskdefinitions?view=graph-rest-beta) using application permissions. This abstract task definition will be used to create task that will hold the job for your application. You need to define at least one task definition per tenant, which can be associated with any number of printers in the tenant using task triggers (see step 4).
+
+2. [Register one or more virtual printers](/graph/api/printer-create?view=graph-rest-beta) using an administrator authentication token and a `null` **physicalDeviceId**. A "virtual printer" is just a printer object in Universal Print without a physical device attached. Usually, users will print to virtual printers and later pick up their print jobs at a physical print device. See step 6.
+
+3. [Update the attributes of your virtual printer](/graph/api/printer-update?view=graph-rest-beta) by using application permissions and an `application/ipp` media type (see examples).
```
Comment (nilakhan) [inline concepts/universal-print-concept-overview.md:72]: Missed a couple of things: 
(see examples): we haven't provided examples in ipp.

File: concepts/universal-print-concept-overview.md [hunk 34]
```diff
@@ -57,6 +57,28 @@ Sharing a printer creates a [printerShare](/graph/api/resources/printershare?vie
...
+
+4. [Create a task trigger for your virtual printer](/graph/api/printer-post-tasktriggers?view=graph-rest-beta) using an administrator authentication token that will associate your task definition with virtual printer.
+
+5. When a print job is submitted to the virtual printer, it will be paused due to the [printTaskTrigger](/graph/api/resources/printtasktrigger?view=graph-rest-beta). A [printTask](/graph/api/resources/printtask?view=graph-rest-beta) with `processing` state will be created based on the associated [printTaskDefinition](/graph/api/resources/printtaskdefinition?view=graph-rest-beta).
+
+6. When the user swipes a badge at a physical printer device, the printer will notify your application. At that time, your application can [fetch the jobs of the associated virtual printer](/graph/api/printer-list-jobs?view=graph-rest-beta) and filter the list to jobs created by the current user.
```
Comment (nilakhan) [inline concepts/universal-print-concept-overview.md:78]: Missed a couple of things:
When the user swipes a badge at a physical printer device, the printer will notify your application. (This is out of scope of UniversalPrint and onus of this is on your application)
